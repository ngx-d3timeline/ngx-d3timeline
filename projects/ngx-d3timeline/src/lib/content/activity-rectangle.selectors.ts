import { createSelector } from '../selector/create-selector';
import { selectTimeScale, selectBandScale } from '../store/timeline-selectors';
import { selectTimeOrientation } from '../store/state';
import { selectNonDraggedActivities } from './activity.selectors';
import {
  getPositionInTimeAxis,
  getPositionInResourceAxis,
  getActivityX,
  getActivityY,
  getActivityTopLeft,
  getRectBreadthInTimeAxis,
  getRectHeight,
  getRectWidth,
  getActivityTransform,
  getActivityRectangle
} from './content-utils';

const selectPositionInTimeAxis = createSelector(selectTimeScale, timeScale =>
  getPositionInTimeAxis.bind(null, timeScale)
);

const selectPositionInResourceAxis = createSelector(
  selectBandScale,
  bandScale => getPositionInResourceAxis.bind(null, bandScale)
);

const selectPositionedActivityY = createSelector(
  selectTimeOrientation,
  selectPositionInResourceAxis,
  selectPositionInTimeAxis,
  (timeOrientation, positionInResourceAxis, positionInTimeAxis) =>
    getActivityY.bind(
      null,
      timeOrientation,
      positionInResourceAxis,
      positionInTimeAxis
    )
);

const selectPositionedActivityX = createSelector(
  selectTimeOrientation,
  selectPositionInResourceAxis,
  selectPositionInTimeAxis,
  (timeOrientation, positionInResourceAxis, positionInTimeAxis) =>
    getActivityX.bind(
      null,
      timeOrientation,
      positionInResourceAxis,
      positionInTimeAxis
    )
);

const selectPositionedActivityTopLeft = createSelector(
  selectPositionedActivityX,
  selectPositionedActivityY,
  (x, y) => getActivityTopLeft.bind(null, x, y)
);

const selectRectBreadthInTimeAxis = createSelector(selectTimeScale, timeScale =>
  getRectBreadthInTimeAxis.bind(null, timeScale)
);

const selectRectBreadthInResourceAxis = createSelector(
  selectBandScale,
  bandScale => bandScale.bandwidth()
);

const selectRectHeight = createSelector(
  selectTimeOrientation,
  selectRectBreadthInTimeAxis,
  selectRectBreadthInResourceAxis,
  (timeOrientation, rectBreadthInTimeAxis, rectBreadthInResourceAxis) =>
    getRectHeight.bind(
      null,
      timeOrientation,
      rectBreadthInTimeAxis,
      rectBreadthInResourceAxis
    )
);

const selectRectWidth = createSelector(
  selectTimeOrientation,
  selectRectBreadthInTimeAxis,
  selectRectBreadthInResourceAxis,
  (timeOrientation, rectBreadthInTimeAxis, rectBreadthInResourceAxis) =>
    getRectWidth.bind(
      null,
      timeOrientation,
      rectBreadthInTimeAxis,
      rectBreadthInResourceAxis
    )
);

const selectPositionedActivityTransform = createSelector(
  selectPositionedActivityTopLeft,
  getTopLeft => getActivityTransform.bind(null, getTopLeft)
);

const selectActivityRectangle = createSelector(
  selectPositionedActivityTransform,
  selectRectWidth,
  selectRectHeight,
  (transform, width, height) =>
    getActivityRectangle.bind(null, transform, width, height)
);

export const selectNonDraggedActivityRectangles = createSelector(
  selectNonDraggedActivities,
  selectActivityRectangle,
  (activities, positionedActivityToRectangle) =>
    activities.map(positionedActivityToRectangle)
);
