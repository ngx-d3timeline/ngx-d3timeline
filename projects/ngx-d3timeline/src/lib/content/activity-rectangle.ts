import { identifier } from '../types';

export interface ActivityRectangle {
  id: identifier;
  title: string;
  showTitle: boolean;
  transform: string;
  width: number;
  height: number;
  fontSize: number;
}
