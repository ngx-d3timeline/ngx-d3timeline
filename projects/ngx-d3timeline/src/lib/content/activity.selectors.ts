import { selectDragEvent, selectPositionedActivities } from '../store/state';
import { createSelector } from '../selector/create-selector';

const selectDragEventId = createSelector(
  selectDragEvent,
  dragEvent => dragEvent && dragEvent.id
);

export const selectNonDraggedActivities = createSelector(
  selectPositionedActivities,
  selectDragEventId,
  (activities, dragEventId) =>
    dragEventId
      ? activities.filter(activity => activity.id !== dragEventId)
      : activities
);
