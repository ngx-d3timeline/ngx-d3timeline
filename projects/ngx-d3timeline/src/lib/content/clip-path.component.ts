import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { Rectangle } from '../core/rectangle';

@Component({
  selector: '[ngx-d3timeline-clip-path]',
  template: `
    <svg:clipPath id="clipRect">
      <svg:rect
        [attr.x]="clipRect.x"
        [attr.y]="clipRect.y"
        [attr.width]="clipRect.width"
        [attr.height]="clipRect.height"
      ></svg:rect>
    </svg:clipPath>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ClipPathComponent {
  @Input() clipRect: Rectangle;
}
