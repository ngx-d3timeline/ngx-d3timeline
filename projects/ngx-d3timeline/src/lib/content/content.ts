export interface EventRectangle {
  id: string | number;
  title: string;
  transform: string;
  width: number;
  height: number;
}
