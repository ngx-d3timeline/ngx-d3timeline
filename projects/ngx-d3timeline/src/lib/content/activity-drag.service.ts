import { Injectable } from '@angular/core';
import { tempStateSelector } from '../store/timeline-selectors';
import { BehaviorSubject } from 'rxjs';
import { Store } from '../store/store';
import { withLatestFrom, filter } from 'rxjs/operators';
import { getDropActivity } from '../drag-utils';
import { drag } from 'd3-drag';
import { select, event } from 'd3-selection';

import * as fromActions from '../store/actions';
import { ActivityRectangle } from './activity-rectangle';

@Injectable({ providedIn: 'root' })
export class ActivityDragService {
  private dragEndSubject = new BehaviorSubject(null);

  drop$ = this.dragEndSubject.pipe(
    filter(() => !!event),
    withLatestFrom(this.store.select(tempStateSelector))
  );

  constructor(private store: Store) {
    this.drop$.subscribe(([, tempState]) => {
      const dropActivity = getDropActivity(
        tempState.bandScale,
        tempState.timeScale,
        tempState.activities,
        tempState.dragEvent,
        tempState.timeOrientation
      );

      this.store.dispatch(
        new fromActions.TimelineDragEndedAction(dropActivity)
      );
    });
  }

  setupDrag(activityRectangle: ActivityRectangle, nativeElement: HTMLElement) {
    const onDrag = drag()
      .on('start', () => this.onDragStarted(activityRectangle))
      .on('drag', () => this.onDragging(activityRectangle))
      .on('end', this.onDragEnded.bind(this));

    onDrag(select(nativeElement));
  }

  private onDragStarted(activityRectangle: ActivityRectangle) {
    this.store.dispatch(
      new fromActions.TimelineDragStartedAction({
        id: activityRectangle.id,
        event
      })
    );
  }

  private onDragging(activityRectangle: ActivityRectangle) {
    this.store.dispatch(new fromActions.TimelineDraggingAction(event));
  }

  private onDragEnded() {
    this.dragEndSubject.next(null);
  }
}
