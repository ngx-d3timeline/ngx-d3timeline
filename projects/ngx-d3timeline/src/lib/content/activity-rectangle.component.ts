import {
  Component,
  Input,
  ElementRef,
  ViewChild,
  AfterViewInit,
  ChangeDetectionStrategy
} from '@angular/core';
import { ActivityRectangle } from './activity-rectangle';
import { ActivityDragService } from './activity-drag.service';

@Component({
  selector: '[ngx-d3timeline-activity-rectangle]',
  template: `
    <svg:g
      class="activity-rectangle"
      [attr.transform]="activityRectangle.transform"
      *ngIf="activityRectangle"
      #activityRectangleEl
    >
      <svg:rect
        [attr.height]="activityRectangle.height"
        [attr.width]="activityRectangle.width"
      ></svg:rect>
      <svg:g *ngIf="showLabel">
        <svg:text dominant-baseline="hanging" dx="2" dy="2">
          {{ activityRectangle.title }}
        </svg:text>
      </svg:g>
    </svg:g>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ActivityRectangleComponent implements AfterViewInit {
  labelFontHeight = 10;
  labelPaddingEachSide = 1;
  minHeightToShowLabel = this.labelFontHeight + this.labelPaddingEachSide * 2;

  @Input() activityRectangle: ActivityRectangle;

  @ViewChild('activityRectangleEl') activityRectangleEl: ElementRef;

  constructor(private activityDragService: ActivityDragService) {}

  ngAfterViewInit() {
    this.setupDrag();
  }

  get showLabel(): boolean {
    return this.activityRectangle.height >= this.minHeightToShowLabel; // TODO consider orientation
  }

  private setupDrag() {
    if (this.activityRectangleEl) {
      this.activityDragService.setupDrag(
        this.activityRectangle.id,
        this.activityRectangleEl.nativeElement
      );
    }
  }
}
