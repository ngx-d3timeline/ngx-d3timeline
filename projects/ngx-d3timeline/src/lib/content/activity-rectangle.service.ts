import { Injectable, ElementRef } from '@angular/core';
import { ActivityDragService } from './activity-drag.service';
import { ActivityRectangle } from './activity-rectangle';

@Injectable({ providedIn: 'root' })
export class ActivityRectangleService {
  constructor(private dragService: ActivityDragService) {}

  setupDrag(
    activityRectangle: ActivityRectangle,
    activityRectangleEl: ElementRef
  ) {
    if (activityRectangleEl) {
      this.dragService.setupDrag(
        activityRectangle.id,
        activityRectangleEl.nativeElement
      );
    }
  }
}
