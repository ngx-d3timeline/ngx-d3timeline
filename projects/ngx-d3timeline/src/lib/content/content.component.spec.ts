import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ContentService } from './content.service';
import { of } from 'rxjs';
import { ContentComponent } from './content.component';
import { Component, Input } from '@angular/core';
import { ActivityRectangle } from '../activity-rectangle/activity-rectangle';
import { Rectangle } from '../core/rectangle';

@Component({
  selector: '[ngx-d3timeline-clip-path]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeClipPathComponent {
  @Input() clipRect: Rectangle;
}

@Component({
  selector: '[ngx-d3timeline-activity-rectangle]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeActivityRectangleComponent {
  @Input() activityRectangle: ActivityRectangle;
}

describe('ContentComponent', () => {
  let fixture: ComponentFixture<ContentComponent>;
  let contentService: ContentService;
  let mockActivityRectangles: ActivityRectangle[];

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        ContentComponent,
        FakeActivityRectangleComponent,
        FakeClipPathComponent
      ],
      providers: [{ provide: ContentService, useValue: jest.fn() }]
    });

    fixture = TestBed.createComponent(ContentComponent);
    contentService = TestBed.inject(ContentService);
    mockActivityRectangles = [
      {
        id: 1,
        title: 'Activity 1',
        transform: 'translate(50,0)',
        type: 'Driving',
        width: 50,
        height: 80,
        fontSize: 10,
        showTitle: true,
        fontFace: 'Arial'
      },
      {
        id: 2,
        title: 'Activity 2',
        transform: 'translate(130,0)',
        type: 'Driving',
        width: 60,
        height: 90,
        fontSize: 10,
        showTitle: true,
        fontFace: 'Arial'
      }
    ];
  });

  describe('activityRectangles null', () => {
    beforeEach(() => {
      contentService.activityRectangles$ = of(null);
    });

    it('should not render', () => {
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should not render when draggingRectangle set', () => {
      contentService.draggingRectangle$ = of(mockActivityRectangles[0]);

      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should not render when dropRectangle set', () => {
      contentService.dropRectangle$ = of(mockActivityRectangles[0]);

      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should not render when fromRectangle set', () => {
      contentService.fromRectangle$ = of(mockActivityRectangles[0]);

      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });

  describe('activityRectangles set', () => {
    it('should render', () => {
      contentService.activityRectangles$ = of(mockActivityRectangles);
      contentService.draggingRectangle$ = of(mockActivityRectangles[0]);
      contentService.dropRectangle$ = of(mockActivityRectangles[0]);
      contentService.fromRectangle$ = of(mockActivityRectangles[0]);
      contentService.clipRect$ = of({ x: 0, y: 0, width: 10, height: 10 });

      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });
});
