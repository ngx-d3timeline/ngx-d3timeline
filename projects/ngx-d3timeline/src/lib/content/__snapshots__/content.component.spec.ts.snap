// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentComponent eventRectangles null should not render 1`] = `
<div
  id="root0"
  ngx-d3timeline-content=""
>
  
</div>
`;

exports[`ContentComponent eventRectangles null should not render when draggingRectangle set 1`] = `
<div
  id="root1"
  ngx-d3timeline-content=""
>
  
</div>
`;

exports[`ContentComponent eventRectangles null should not render when dropRectangle set 1`] = `
<div
  id="root2"
  ngx-d3timeline-content=""
>
  
</div>
`;

exports[`ContentComponent eventRectangles set should render 1`] = `
<div
  id="root3"
  ngx-d3timeline-content=""
>
  
  <g
    class="content-group"
  >
    
    <g
      ngx-d3timeline-event-rectangle=""
    >
      <g />
    </g>
    <g
      ngx-d3timeline-event-rectangle=""
    >
      <g />
    </g>
    
    <g
      class="drop-event-rectangle"
      ngx-d3timeline-event-rectangle=""
    >
      <g />
    </g>
    
    <g
      ngx-d3timeline-event-rectangle=""
    >
      <g />
    </g>
  </g>
</div>
`;
